src/libraries/System.Text.Encoding.CodePages/src/Data/Tools/EncodingDataGenerator.cs

```diff
@@ -0,0 +1,466 @@
+                    if (names.Equals(existing))
+                    {
+                        Console.WriteLine("Code page names {0} for code page {1} in file {2} at line {3} is a duplicate entry, and can be removed.", names, codepage, path, line.Key);
+                    }
+                    else
+                    {
+                        Console.WriteLine("Code page {0} in file {1} at line {2} is mapped to multiple names; new is {3}, old was {4}", codepage, path, line.Key, names, existing);
+                    }
+                }
+                else
+                {
+                    preferredNames[codepage] = names;
+                }
+            }
+
+            return preferredNames;
+        }
+
+        private IEnumerable<KeyValuePair<int, string[]>> DelimitedFileRows(string path, int columns = 0)
+        {
+            using (var input = File.OpenText(path))
+            {
+                int lineNumber = 1;
+                string line;
+
+                for (; (line = input.ReadLine()) != null; ++lineNumber)
+                {
+                    if (line.StartsWith(CommentIndicator) || string.IsNullOrWhiteSpace(line))
+                    {
+                        continue;
+                    }
+
+                    string[] values = line.Split(FieldDelimiter);
+
+                    if (columns > 0 && values.Length != columns)
+                    {
+                        Console.WriteLine("Parsing mapping in file {0}, line {1}.  Expected {2} fields, saw {3}: {4}", path, lineNumber, columns, values.Length, line);
+                    }
+
+                    yield return KeyValuePair.Create(lineNumber, line.Split(FieldDelimiter));
+                }
+            }
+        }
+
+        private const string Header =
+
+@"// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+//
+// THIS IS AN AUTOGENERATED FILE
+// IT IS GENERATED FROM EncodingDataGenerator Tool.
+//
+//     EncodingDataGenerator.exe  --IanaMapping ..\..\Data\CodePageNameMappings.csv --PreferredIANANames ..\..\Data\PreferredCodePageNames.csv --Output EncodingTable.Data.cs
+//
+
+namespace {2}
+{{
+    internal static partial class {3}
+    {{
+";
+
+        // The format is:
+        //     0 - IANA name
+        //     1 - codepage
+        // Ordered by alphabetized IANA name
+        private const string EncodingNames =
+@"
+        // s_encodingNames is the concatenation of all supported IANA names for each codepage.
+        // This is done rather than using a large readonly array of strings to avoid
+        // generating a large amount of code in the static constructor.
+        // Using indices from s_encodingNamesIndices, we binary search this string when mapping
+        // an encoding name to a codepage. Note that these names are all lowercase and are
+        // sorted alphabetically.
+        private const string s_encodingNames =|
```

Comment:
Would it be more appropriate to make it a [Source Generator](https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/) project, instead of a standalone console app?

Reply:
I don't think we need to do that. I am not expecting this will used much. I just added it to ensure we have the tool code that we can use if we ever needed it.